pykafka/producer.py

```diff
@@ -180,56 +331,128 @@ def _get_partition_msgs(partition_id, req):
             attempt += 1
             if attempt < self._max_retries:
                 time.sleep(self._retry_backoff_ms / 1000)
-                self._produce(to_retry, attempt)
+                # we have to call _produce here since the broker/partition
+                # target for each message may have changed
+                for _ in self._produce(to_retry, attempt):
+                    pass
             else:
                 raise ProduceFailureError('Unable to produce messages. See log for details.')
 
-    def _partition_messages(self, messages):
-        """Assign messages to partitions using the partitioner.
+    @raise_worker_exceptions
+    def _update_leaders(self):
+        """Ensure each message in each queue is in the queue owned by its
+            partition's leader
 
-        :param messages: Iterable of messages to publish.
-        :returns:        Generator of ((key, value), partition_id)
+        This function empties all broker queues, maps their messages to the
+        current leaders for their partitions, and enqueues the messages in
+        the appropriate queues.
         """
-        partitions = self._topic.partitions.values()
-        for message in messages:
-            if isinstance(message, basestring):
-                key = None
-                value = message
-            else:
-                key, value = message
-            value = str(value)
-            yield (key, value), self._partitioner(partitions, message).id
+        # empty queues and figure out updated partition leaders
+        new_queue_contents = defaultdict(list)
+        for owned_broker in self._owned_brokers.itervalues():
+            owned_broker.lock.acquire()
```

Comment:
My guess would be that's not possible here, because you need to hold locks across all brokers, to prevent new messages getting slipped-in between the old ones while they are being redistributed.